### PR TITLE
Feature: BERT integration

### PR DESCRIPTION
**Pull request by 3ripleM**

### CODE REVIEW

**Comment by 3ripleM** [comment]
@hosseinfani, I need access to BERT Repository in fanilab to push my codes there.
right now the codes are in my github repository.

File: src/aml/bert.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    # Due to model cannot handle long sentences, we need to filter out long sentences
+    REVIEW_MAX_LENGTH = 511
+
+    for r in original_reviews:
+        if not len(r.aos[0]):
+            continue
+        else:
+            aos_list = []
+
+            for aos_instance in r.aos[0]:
```
Comment (hosseinfani) [inline src/aml/bert.py:39]: @3ripleM pls make it one line

File: src/aml/bert.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+            for idx, word in enumerate(r.sentences[0]):
+                if idx in aos_list:
+                    tag = word + '=T-POS' + ' '
+                    text += tag
+                else:
+                    tag = word + '=O' + ' '
+                    text += tag
+
+            if len(text.rstrip()) > REVIEW_MAX_LENGTH:
```
Comment (hosseinfani) [inline src/aml/bert.py:52]: @3ripleM one line pls

File: src/aml/bert.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                continue
+
+            reviews_list.append(text.rstrip())
+
+            aos_list_per_review = []
+
+            for idx, word in enumerate(r.sentences[0]):
+                if idx in aos_list:
```
Comment (hosseinfani) [inline src/aml/bert.py:60]: @3ripleM same

File: src/aml/bert.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            label_list.append(aos_list_per_review)
+
+    return reviews_list, label_list
+
+
+def save_train_reviews_to_file(original_reviews: List[Review], output: str) -> None:
+    train, _ = convert_reviews_from_lady(original_reviews)
+    dev, _ = convert_reviews_from_lady(original_reviews)
+
+    if not os.path.isdir(output):
```
Comment (hosseinfani) [inline src/aml/bert.py:72]: @3ripleM

File: src/aml/bert.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+#   year         = {2013},
+#   url          = {https://doi.org/10.1145/2488388.2488514},
+#   biburl       = {https://dblp.org/rec/conf/www/YanGLC13.bib},
+# }
+class BERT(AbstractAspectModel):
+    def __init__(self, naspects, nwords): 
+        super().__init__(naspects, nwords)
+
+    # TODO: What is this? should I change it?
+    def load(self, path):
```
Comment (hosseinfani) [inline src/aml/bert.py:141]: @3ripleM this is for test purpose as we only need to load the model if it's already there. if you don't need and use the baseline code for this, as I see blow, simply remove it.

File: src/aml/bert.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            raise RuntimeError(f'Error in training BERT model: {e}')
+
+    def infer_batch(self, reviews_test, h_ratio, doctype, output):
+        save_test_reviews_to_file(reviews_test, h_ratio, output)
+
+        absa_home = 'CHANGE_ME'
+        args = {
+            # TODO: change this to the model directory
+            'absa_home': f'{absa_home}',
+            'ckpt': f'{absa_home}/checkpoint-1200',
```
Comment (hosseinfani) [inline src/aml/bert.py:193]: @3ripleM are we going to fix this?

File: src/aml/ctm.py [hunk 7]
```diff
@@ -80,12 +81,17 @@ def get_aspect_words(self, aspect_id, nwords): return self.mdl.get_word_distribu
...
     def infer_batch(self, reviews_test, h_ratio, doctype):
         reviews_test_ = []
-        reviews_aspects = []
+        reviews_aspects: List[List[List[str]]] = []
         for r in reviews_test:
             r_aspects = [[w for a, o, s in sent for w in a] for sent in r.get_aos()]  # [['service', 'food'], ['service'], ...]
-            if len(r_aspects[0]) == 0: continue  # ??
-            if random.random() < h_ratio: r_ = r.hide_aspects()
-            else: r_ = r
+
```
Comment (hosseinfani) [inline src/aml/ctm.py:87]: @3ripleM seems no change. pls only push the required lines of changes. also one line pls

File: src/aml/ctm.py [hunk 8]
```diff
@@ -95,9 +101,10 @@ def infer_batch(self, reviews_test, h_ratio, doctype):
         processed, unprocessed, vocab, _ = WhiteSpacePreprocessingStopwords(corpus_test, stopwords_list=[]).preprocess()
         testing_dataset = self.tp.transform(text_for_contextual=unprocessed, text_for_bow=processed)
         reviews_pred_aspects = self.mdl.get_doc_topic_distribution(testing_dataset, n_samples=self.nsamples)
-        pairs = []
+        pairs: List[Tuple[List[str], ]] = []
         for i, r_pred_aspects in enumerate(reviews_pred_aspects):
             r_pred_aspects = [[(j, v) for j, v in enumerate(r_pred_aspects)]]
             pairs.extend(list(zip(reviews_aspects[i], self.merge_aspects_words(r_pred_aspects, self.nwords))))
 
+        print(pairs)
```
Comment (hosseinfani) [inline src/aml/ctm.py:109]: @3ripleM don we need this?

File: src/aml/lda.py [hunk 9]
```diff
@@ -73,7 +73,10 @@ def get_aspect_words(self, aspect_id, nwords): return self.mdl.show_topic(aspect
     def infer(self, review, doctype):
         review_aspects = []
         review_, _ = super(Lda, self).preprocess(doctype, [review])
-        for r in review_: review_aspects.append(self.mdl.get_document_topics(self.dict.doc2bow(r), minimum_probability=self.mdl.minimum_probability))
+        for r in review_:
```
Comment (hosseinfani) [inline src/aml/lda.py:76]: @3ripleM same comment

File: src/aml/mdl.py [hunk 10]
```diff
@@ -27,23 +30,31 @@ def quality(self, metric):
...
+    def infer(self, review: Review, doctype: str): pass
+    def infer_batch(self, reviews_test: List[Review], h_ratio: int, doctype: str, output: str):
         pairs = []
+
         for r in reviews_test:
             r_aspects = [[w for a, o, s in sent for w in a] for sent in r.get_aos()]  # [['service', 'food'], ['service'], ...]
-            if len(r_aspects[0]) == 0: continue  # ??
-            if random.random() < h_ratio: r_ = r.hide_aspects()
-            else: r_ = r
+
```
Comment (hosseinfani) [inline src/aml/mdl.py:39]: same comment

File: src/main.py [hunk 11]
```diff
@@ -1,16 +1,28 @@
-import argparse, os, json, time
+import argparse
+import os
+import json
+import time
+from typing import List, Tuple
 from tqdm import tqdm
-import numpy as np, pandas as pd
+import numpy as np
```
Comment (hosseinfani) [inline src/main.py:7]: @3ripleM the general idea is not to add vertical lines :)

File: src/main.py [hunk 12]
```diff
@@ -118,15 +131,66 @@ def test(am, test, f, output):
...
         am.load(f'{output}/f{f}.')
         print('3.3. Testing aspect model ...')
-        pairs = am.infer_batch(reviews_test=test, h_ratio=params.settings['test']['h_ratio'], doctype=params.settings['prep']['doctype'])
+        pairs = am.infer_batch(reviews_test=test, h_ratio=params.settings['test']['h_ratio'], doctype=params.settings['prep']['doctype'], output=output)
         pd.to_pickle(pairs, f'{output}f{f}.model.pred.{params.settings["test"]["h_ratio"]}')
 
-def evaluate(input, output):
+
+# TODO: Refactor this to convert Bert pairs to LADy pairs
+def evaluate_bert(input: str, output: str):
```
Comment (hosseinfani) [inline src/main.py:139]: @3ripleM we should not have this here. the main should not be customized for a specific model

File: src/params.py [hunk 13]
```diff
@@ -32,6 +32,7 @@ def to_range(range_str): return range(int(range_str.split(':')[0]), int(range_st
                 # {'no_below': 10,   # happen less than no_below number in total
                 #  'no_above': 0.9}  # happen in no_above percent of reviews
         'rnd': {},
+        'bert': {},
```
Comment (hosseinfani) [inline src/params.py:35]: @3ripleM does it have any hypeparameter??

**Review by hosseinfani — CHANGES_REQUESTED** [review]

File: src/aml/bert.py (hunk 1 above)
Comment (hosseinfani) [reply src/aml/bert.py:39]: @3ripleM 
I respect your opinion and it's true that **long** lines become confusing. However, we try to make it one line as far as they're not long :) pls follow our lab's convention. Thank you.

File: src/main.py (hunk 12 above)
Comment (3ripleM) [reply src/main.py:139]: Thats correct and I totally agree with you. this is a monkey patch for the purpose of the paper then ma and farniam or you can make it cleaner (we should just write a function to convert bert pairs to our pairs)

Comment (hosseinfani) [reply src/main.py:139]: put it sw inside the bert class then

File: src/main.py (hunk 11 above)
Comment (3ripleM) [reply src/main.py:7]: This code style makes the codes harder to read. it is not bad to write a lot of code. it is bad that the code is not readable.
I know that python developers can understand that but they need to focus on understanding what is imported from where and also what is the aliases. 

This is a report on UX development and why long lines are bad for readability.
https://baymard.com/blog/line-length-readability

File: src/aml/bert.py (hunk 6 above)
Comment (3ripleM) [reply src/aml/bert.py:193]: @hosseinfani  Sure, we are not done yet. I just made a draft PR for followups

File: src/aml/ctm.py (hunk 8 above)
Comment (3ripleM) [reply src/aml/ctm.py:109]: no we do not.

File: src/aml/bert.py (hunk 1 above)
Comment (3ripleM) [reply src/aml/bert.py:39]: @hosseinfani 
Whenever the characters exceed 80 chars they are long (and it is not only my opinion).

File: src/main.py (hunk 11 above)
Comment (hosseinfani) [reply src/main.py:7]: @3ripleM 
you're right. but for trivial packages that are imported in almost all ml project like numpy and the likes, all in this domain have reached to a standard alias like np or pd ... 

however, for those that are new or are from our own architecture, we follow what you suggested.

**Comment by hosseinfani** [comment]
@3ripleM 
> @hosseinfani, I need access to BERT Repository in fanilab to push my codes there. right now the codes are in my github repository.

done!

**Review by hosseinfani — APPROVED** [review]

**Pull request merged**